File: admin-ui/src/app/response/commodity.ts
import { Base, Page } from './response';
import { Tag } from './tag';

export interface CommodityMenu {
    id: number;
    title: string;
    parentId: number;
    children: number;
    comment: string;
}

export interface CommodityMenuListResponse extends Base {
    list: Array<CommodityMenu>;
}

export interface CommoditySpec {
    id: number;
    title: string;
    commodityId: number;
    breed: string;
    origin: string;
    feature: string;
    spec: string;
    store: string;
    productDatetime: string;
    price: number;
    inventory: number;
}

export interface CommoditySpecListResponse extends Page {
    list: Array<CommoditySpec>;
}


export interface Commodity {
    id: number;
    title: string;
    summary: string;
    comment?: string;
    description: string;
    status: number;
    ranking: number;
    rankingCount: number;
    viewCount: number;
    shareCount: number;
    createAt?: Date;
    updateAt?: Date;
    tags?: Array<Tag>;
    specifications?: Array<CommoditySpec>;
    menus?: Array<CommodityMenu>;
}

export interface CommodityListResponse extends Page {
    list: Array<Commodity>;
}

File: src/main/java/com/iprzd/zshop/entity/commodity/Commodity.java
package com.iprzd.zshop.entity.commodity;

import com.iprzd.zshop.entity.Tag;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

import javax.persistence.*;
import java.util.Date;
import java.util.List;

@Entity
@Table(name = "t_commodity")
@Getter
@Setter
@NoArgsConstructor
@ToString
@EqualsAndHashCode
public class Commodity {

    @Id @GeneratedValue
    private long id;
    private String title;
    private String summary;
    private String image;
    private String comment;
    @Column(columnDefinition="text")
    private String description;
    @Column(columnDefinition="tinyint default 0")
    private int status;
    @Column(columnDefinition="int default 10")
    private int ranking;
    @Column(columnDefinition="int default 1")
    private int rankingCount;
    @Column(columnDefinition="int default 0")
    private int viewCount;
    @Column(columnDefinition="int default 0")
    private int shareCount;
    private Date createAt;
    private Date updateAt;

    @ManyToMany(cascade = CascadeType.PERSIST, fetch = FetchType.LAZY)
    @JoinTable(name = "t_commodity_tag",
            joinColumns = @JoinColumn(name = "commodity_id", referencedColumnName = "id"),
            inverseJoinColumns = @JoinColumn(name = "tag_id", referencedColumnName = "id"))
    private List<Tag> tags;

    @ManyToMany(cascade = CascadeType.PERSIST, fetch = FetchType.LAZY)
    @JoinTable(name = "t_commodity_spec_map",
            joinColumns = @JoinColumn(name = "commodity_id", referencedColumnName = "id"),
            inverseJoinColumns = @JoinColumn(name = "spec_id", referencedColumnName = "id"))
    private List<Specification> specifications;

    // CascadeType.PERSIST, CascadeType.REFRESH, CascadeType.MERGE CascadeType. REMOVE CascadeType. REMOVE
    @ManyToMany(cascade =  { CascadeType.PERSIST }, fetch = FetchType.LAZY)
    @JoinTable(name = "t_commodity_menu_map",
            joinColumns = @JoinColumn(name = "commodity_id", referencedColumnName = "id"),
            inverseJoinColumns = @JoinColumn(name = "menu_id", referencedColumnName = "id"))
    @OrderBy(value = "id asc")
    private List<Menu> menus;
}
